docs/alternativesemantics.md

```diff
@@ -0,0 +1,202 @@
+     |  tau ~>_r tau  |  tau <~_r tau  |  X
+```
+where `I` are interfaces
+   `r` are region variables
+   `X` are type variables
+
+We restrict types on fields and generic parameters to not mention `r`, i.e. they cannot contain `r`, `tau ~>_r tau` and `tau <~_r tau`.
+
+**[Sylvan is this too restrictive]**
+
+We interpret the semantics of types with 
+```
+[[ _ ]] : 
+    (region var -> region) 
+        -> (type var -> P(C, region, cap)) 
+            -> P(C, region, cap)
+```
+where
+```
+  [[ iso ]](R,T) = { C,r,iso | C \in Class, r \in region}
+  [[ r ]](R,T) = { C,R(r),cap | C \in Class, cap \in Caps }
+  [[ C ]](R,T) = { C,r,cap | r \in region, cap \in Caps }
+  [[ X ]](R,T) = T(X)
+  [[ tau ~>_r tau ]](R, T) = 
+    { V1 ~>_R(r) V2 | V1 \in [[tau1]](R,T) /\ V2 \in [[tau2]](R,T)}
+  [[ tau <~_r tau ]](R, T) = 
+    { V1 <~_R(r) V2 | V1 \in [[tau1]](R,T) /\ V2 \in [[tau2]](R,T)}
+```
+
+Aliasing viewpoint adaptation
+```
+C1,r1,mut ~>_r C2,_,mut = C2,r1,mut
+C1,r1,iso ~>_r C2,_,mut = C2,r1,mut
+C1,r1,imm ~>_r C2,_,mut = C2,r1,imm
+C1,r1,mut ~>_r C2,_,iso = C2,r,mut
+C1,r1,iso ~>_r C2,_,iso = C2,r,mut
+C1,r1,imm ~>_r C2,_,iso = C2,r1,imm
+C1,r1,mut ~>_r C2,_,imm = C2,r1,imm
+C1,r1,iso ~>_r C2,_,imm = C2,r1,imm
+C1,r1,imm ~>_r C2,_,imm = C2,r1,imm
+```
+**[Could be a relational definition to allow for unrestricted r on the right hand side.]**
+
+Extracting viewpoint adaptation
+```
+C1,r1,mut <~_r C2,_,mut = C2,r1,mut
+C1,r1,iso <~_r C2,_,mut = C2,r1,mut
+C1,r1,imm <~_r C2,_,mut = C2,r1,imm
+C1,r1,mut <~_r C2,_,iso = C2,r,iso
+C1,r1,iso <~_r C2,_,iso = C2,r,iso
+C1,r1,imm <~_r C2,_,iso = C2,r1,imm
+C1,r1,mut <~_r C2,_,imm = C2,r1,imm
+C1,r1,iso <~_r C2,_,imm = C2,r1,imm
+C1,r1,imm <~_r C2,_,imm = C2,r1,imm
+```
+
+The most important difference between extracting and aliasing
+viewpoint adaption is how they behave on `iso` fields. Extracting provides an `iso` in the supplied region, and aliasing provides a `mut` in the supplied region.
+
+The definitions take a region parameter so that "freshness" can be suitably defined. 
+
+### Type Rules
+
+Aliasing assignment generates a fresh region
+if it follows an `iso`, if it follows an `imm` then it is in the immutable region, otherwise it is in the same region. 
+
+Read versus Write types for a field
+```
+   field_r(tau1 | tau2, f) = field_r(tau1, f) | field_r(tau2, f)
+   field_r(tau1 & tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+   field_w(tau1 | tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+   field_w(tau1 & tau2, f) = field_r(tau1, f) & field_r(tau2, f)
+```
+]
+
+```
+field_R(f, tau_x) = tau_f
+fresh r
+-------------------------------------
+G |- z = x.f -| G, z: tau_x ->r tau_f
+```
+
+Extraction uses a fresh region
+```
+G(x) : tau_x
+G(y) : tau_y
+field_w(f, tau_x) = tau_f_w
+field_r(f, tau_x) = tau_f_r
+exists r'. tau_y <: tau_x <~r' tau_f_w
+fresh r
+--------------------------------------
+G |- z = (x.f = y) -| G, z: tau_x <~r tau_f_r
+```

```

Comment:
I don't think this is enforcing that `y` is in the same region as `x` correctly.

If we have no region restrictions on either `y` or `x`, then if subtyping is subset with the semantics above, then it will be allowed, but shouldn't be.